Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/events/server/qrcodepostertemplate/QrCodePosterTemplateServer.kt

```diff
@@ -1,60 +1,67 @@
-        Timber.d("Received: %s", response)
-
-        if (!response.isSuccessful) {
-            // TODO return cached or default response
-            throw HttpException(response)
-        }
-        if (response.body() == null) {
-            throw IllegalStateException("Response is successful, but body is empty.")
+        return try {
+            QrCodePosterTemplate.QRCodePosterTemplateAndroid.parseFrom(binaryTemplate)
+        } catch (exception: InvalidProtocolBufferException) {
+            throw QrCodePosterTemplateInvalidResponseException(
+                message = "QR Code poster template could not be parsed",
+                cause = exception
+            )
         }
+    }
 
-        val fileMap = response.body()!!.byteStream().unzip().readIntoMap()
-
-        val exportBinary = fileMap[EXPORT_BINARY_FILE_NAME]
-        val exportSignature = fileMap[EXPORT_SIGNATURE_FILE_NAME]
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    suspend fun getTemplateFromApiOrCache(): ByteArray {
+        return try {
+            val response = api.getQrCodePosterTemplate()
+            if (response.body() == null) {
+                throw IllegalStateException("Response is successful, but body is empty.")
+            }
 
-        if (exportBinary == null || exportSignature == null) {
-            throw QrCodePosterTemplateInvalidResponseException(message = "Unknown files: ${fileMap.keys}")
-        }
+            val fileMap = response.body()!!.byteStream().unzip().readIntoMap()
 
-        val hasValidSignature = signatureValidation.hasValidSignature(
-            exportBinary,
-            SignatureValidation.parseTEKStyleSignature(exportSignature)
-        )
+            val exportBinary = fileMap[EXPORT_BINARY_FILE_NAME]
+            val exportSignature = fileMap[EXPORT_SIGNATURE_FILE_NAME]
 
-        if (!hasValidSignature) {
-            throw QrCodePosterTemplateInvalidResponseException(message = "Invalid Signature!")
-        }
+            if (exportBinary == null || exportSignature == null) {
+                throw QrCodePosterTemplateInvalidResponseException(message = "Unknown files: ${fileMap.keys}")
+            }
 
-        return try {
-            QrCodePosterTemplate.QRCodePosterTemplateAndroid.parseFrom(exportBinary)
-        } catch (exception: InvalidProtocolBufferException) {
-            throw QrCodePosterTemplateInvalidResponseException(
-                message = "QR Code poster template could not be parsed",
-                cause = exception
+            val hasValidSignature = signatureValidation.hasValidSignature(
+                exportBinary,
+                SignatureValidation.parseTEKStyleSignature(exportSignature)
             )
+
+            if (!hasValidSignature) {
+                throw QrCodePosterTemplateInvalidResponseException(message = "Invalid Signature!")
+            }
+
+            templateCache.saveTemplate(exportBinary)
+            exportBinary
```

Comment:
Could you please do me a favour and return the file directly along with other response data?
Currently I need to convert those bytes again into a temporary file and delete afterwards. It would be great if you can return a response object such as:
```kotlin
TemplateResponse(
   templateFile : File, // File path in cache directory
   offsetX : Float,
   offsetY : Float,
  ...
)
```

Reply:
You shouldn't call this method directly, but instead call `downloadQrCodePosterTemplate()` which returns the protobuf that contains these values.

Reply:
Still I have to convert the PDF bytes into file right? if you have the file already cached, It would be nice to return the path directly

Reply:
Ah, now I get it. Yes makes sense.

Reply:
but in case I remove my `QrCodePosterTemplateCache` implementation and use only okhttp cache, as you suggested in the previous comment, we wouldn't have direct access to the file anyway.

Reply:
👍🏼